### PR TITLE
Fix: wso2/product-ei#2856

### PR DESCRIPTION
**Pull request by SangeethKarunaratne**
Introduced forcedDefault attribute in QueryParam to set default
values of params in GET request.

Fixes: https://github.com/wso2/product-ei/issues/2856

## Purpose
Set default values of input parameters when invoking a REST resource  with GET method when parameters are not passed.

eg:

```
<query id="GetEmployeeDetails" useConfig="Datasource">
      <sql>select EmployeeNumber, FirstName, LastName, Email from Employees 
       where EmployeeNumber=:EmployeeNumber and FirstName=:FirstName</sql>
      <result element="Employees" rowName="Employee">
         <element column="EmployeeNumber" name="EmployeeNumber" xsdType="string"/>
         <element column="FirstName" name="FirstName" xsdType="string"/>
         <element column="LastName" name="LastName" xsdType="string"/>
         <element column="Email" name="Email" xsdType="string"/>
      </result>
      <param defaultValue="1" name="EmployeeNumber" sqlType="STRING"/>
      <param name="FirstName" sqlType="STRING"/>
   </query>


<resource method="GET" path="Employee/{FirstName}">
      <call-query href="GetEmployeeDetails">
         <with-param name="EmployeeNumber" query-param="EmployeeNumber"/>
         <with-param name="FirstName" query-param="FirstName"/>
      </call-query>
   </resource>
```
http://localhost:8280/services/RDBMSDataService.HTTPEndpoint/Employee/Smith

When the above request is made the FirstName - Smith is only passed as query params but when the message is received the value for EmployeeNumber is set as "null' even  though there is no input for EmployeeNumber.

## Goals
To set default values when a GET request is made upon a resource.
## Approach

`  <param defaultValue="1"  forceDefault="true" name="EmployeeNumber" sqlType="STRING"/>
`

Introduced attribute 'forceDefault" that can be inserted  to any query-param. The default value for forceDefault is false.
When forceDefault="False" null values can be passed and when forceDefault="True" null values are being replaced by default values if default values are given else "null" is passed not the default values.

### CODE REVIEW

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/description/query/QueryFactory.java [hunk 1]
```diff
@@ -1096,13 +1097,16 @@ private static OutputElementGroup createOutputElementGroup(DataService dataServi
 			if (paramType == null || paramType.trim().length() == 0) {
 				paramType = QueryParamTypes.SCALAR;
 			}
+			if (paramEl.getAttributeValue(new QName(DBSFields.FORCEDEFAULT)) != null) {
+			    forceDefault = Boolean.parseBoolean(paramEl.getAttributeValue(new QName(DBSFields.FORCEDEFAULT)));
```
Comment (dilee) [inline components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/description/query/QueryFactory.java]: Let's break this line into two

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/engine/CallQuery.java [hunk 2]
```diff
@@ -315,7 +315,13 @@ protected void executeElement(XMLStreamWriter xmlWriter, ExternalParamCollection
 				paramObj = params.getParam(paramName);
 			}
 			if (paramObj != null) {
-			    qparams.put(withParam.getName(), paramObj.getValue());
+				if (paramObj.getValue().getScalarValue() == null
+					&& params.getTempEntries().containsKey(withParam.getName())
+					&& queryParamMap.get(paramName).isForceDefault()) {
+					continue;
```
Comment (chanikag) [inline components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/engine/CallQuery.java]: Let's put a comment explaining why we continue at this condition?

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/WSDLToDataService.java [hunk 3]
```diff
@@ -360,7 +360,7 @@ public static void printDOM(Document dom) {
 					prop.isArray() ? DBConstants.QueryParamTypes.ARRAY : 
 						DBConstants.QueryParamTypes.SCALAR, 
 					i + 1, // ordinal
-					null, null, new ArrayList<Validator>()));
+					null, null, new ArrayList<Validator>(),false));
```
Comment (chanikag) [inline components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/WSDLToDataService.java]: Code formatting issue

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/description/query/QueryFactory.java [hunk 4]
```diff
@@ -1096,13 +1097,16 @@ private static OutputElementGroup createOutputElementGroup(DataService dataServi
...
+			}
 			/* retrieve validators */
 			List<Validator> validators = getValidators(paramType, paramEl);
             /* retrieve struct type  */
             structType = paramEl.getAttributeValue(new QName(DBSFields.STRUCT_TYPE));
 			queryParams.add(new QueryParam(name, sqlType, type, paramType, ordinal,
                     defaultValue == null ? null : new ParamValue(defaultValue), structType,
-                    validators));
+                    validators,forceDefault));
```
Comment (chanikag) [inline components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/description/query/QueryFactory.java]: Formatting issue

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/engine/CallQuery.java [hunk 5]
```diff
@@ -315,7 +315,13 @@ protected void executeElement(XMLStreamWriter xmlWriter, ExternalParamCollection
 				paramObj = params.getParam(paramName);
 			}
 			if (paramObj != null) {
-			    qparams.put(withParam.getName(), paramObj.getValue());
+				if (paramObj.getValue().getScalarValue() == null
```
Comment (chanikag) [inline components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/engine/CallQuery.java]: Code is not formatted? (indenting)

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/engine/QueryParam.java [hunk 6]
```diff
@@ -59,10 +59,12 @@
     private List<Validator> validators;
 
     private String structType;
+
+    private boolean forceDefault;
 	
 	public QueryParam(String name, String sqlType, String type, String paramType, 
 			int ordinal, ParamValue defaultValue, String structType,
-			List<Validator> validators) throws DataServiceFault {
+			List<Validator> validators,boolean forceDefault) throws DataServiceFault {
```
Comment (chanikag) [inline components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/engine/QueryParam.java]: Code formatting issue. Need a space after ,

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/engine/QueryParam.java [hunk 7]
```diff
@@ -159,8 +162,12 @@ public ParamValue getDefaultValue() {
     public String getStructType() {
         return structType;
     }
-    
-    public boolean hasDefaultValue() {
+
+	public boolean isForceDefault() {
```
Comment (chanikag) [inline components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/engine/QueryParam.java]: Code indentation issue

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/script/DSGenerator.java [hunk 8]
```diff
@@ -851,7 +851,7 @@ private Query getSelectWithKeyQuery(List<String> param,
 			}
 			List<Validator> validator = new ArrayList<Validator>();
 			QueryParam queryParam = new QueryParam(pName, sqlType, DBConstants.DataServiceGenerator.IN,
-					DBConstants.DataServiceGenerator.SCALAR, ordinal, null, null, validator);
+					DBConstants.DataServiceGenerator.SCALAR, ordinal, null, null, validator,false);
```
Comment (chanikag) [inline components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/script/DSGenerator.java]: Need a space after ,

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/engine/CallQuery.java (hunk 2 above)
Comment (SangeethKarunaratne) [reply components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/engine/CallQuery.java]: Fixed

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/WSDLToDataService.java (hunk 3 above)
Comment (SangeethKarunaratne) [reply components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/WSDLToDataService.java]: Fixed

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/description/query/QueryFactory.java (hunk 4 above)
Comment (SangeethKarunaratne) [reply components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/description/query/QueryFactory.java]: Fixed

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/engine/QueryParam.java (hunk 6 above)
Comment (SangeethKarunaratne) [reply components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/engine/QueryParam.java]: Fixed

File: components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/script/DSGenerator.java (hunk 8 above)
Comment (SangeethKarunaratne) [reply components/data-services/org.wso2.carbon.dataservices.core/src/main/java/org/wso2/carbon/dataservices/core/script/DSGenerator.java]: Fixed

File: components/data-services/org.wso2.carbon.dataservices.common/src/main/java/org/wso2/carbon/dataservices/common/DBConstants.java [hunk 9]
```diff
@@ -857,6 +857,7 @@ private DBSFields() {
         public static final String KEY_COLUMNS = "keyColumns";
         public static final String EXPORT = "export";
         public static final String OPTIONAL = "optional";
+        public static final String FORCEDEFAULT = "forceDefault";
```
[the remark this answers is not in the document]
Comment (SangeethKarunaratne) [reply components/data-services/org.wso2.carbon.dataservices.common/src/main/java/org/wso2/carbon/dataservices/common/DBConstants.java]: Fixed

**Pull request merged**